auth/nonce/src/main/java/iu/auth/nonce/OneTimeNumber.java

```diff
@@ -0,0 +1,174 @@
+package iu.auth.nonce;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import edu.iu.IdGenerator;
+import edu.iu.auth.IuAuthorizationChallenge;
+import edu.iu.auth.IuOneTimeNumber;
+import edu.iu.auth.IuOneTimeNumberConfig;
+
+/**
+ * {@link IuOneTimeNumber} implementation class.
+ */
+final class OneTimeNumber implements IuOneTimeNumber, AutoCloseable {
+
+	private static class ActiveCount {
+		private volatile int count;
+	}
+
+	private final Logger LOG = Logger.getLogger(OneTimeNumber.class.getName());
+
+	private IuOneTimeNumberConfig config;
+	private Map<String, IuAuthorizationChallenge> challenge;
+	private Timer purge;
+
+	private int maxConcurrency = 5;
+	private transient Map<String, ActiveCount> activeCount = new HashMap<>();
+
+	/**
+	 * Constructor
+	 * 
+	 * @param config {@link IuOneTimeNumber}
+	 */
+	OneTimeNumber(IuOneTimeNumberConfig config) {
+		this.config = config;
+		challenge = new ConcurrentHashMap<>();
+
+		purge = new Timer();
+		purge.schedule(new TimerTask() {
+			@Override
+			public void run() {
+				final var i = challenge.keySet().iterator();
+				while (i.hasNext())
+					try {
+						IdGenerator.verifyId(i.next(), config.getTimeToLive().toMillis());
+					} catch (Throwable e) {
+						i.remove();
+					}
+			}
+		}, 500L, 500L);
+
+		config.subscribe(this::receive);
+	}
+
+	@Override
+	public String create(String remoteAddress, String userAgent) {
+		if (challenge == null)
+			return IdGenerator.generateId();
+
+		ActiveCount activeCount;
+		synchronized (this.activeCount) {
+			activeCount = this.activeCount.compute(remoteAddress, (k, v) -> v == null ? new ActiveCount() : v);
+			synchronized (activeCount) {
+				activeCount.count++;
+			}
+		}
+
+		try {
+			if (activeCount.count > maxConcurrency) {
+				LOG.info("Blocking excessive nonce generation; " + remoteAddress);
+				return IdGenerator.generateId();
+			}
+
+			try {
+				final var challenge = new IssuedChallenge(remoteAddress, userAgent);

```

Comment:
Could we rename challenge to issuedChallenge to avoid confusion with challenge class field ?